room/compiler-processing/src/main/java/androidx/room/compiler/processing/ksp/KspType.kt

```diff
@@ -0,0 +1,174 @@
+            ?: resolved?.arguments
+            ?: emptyList()
+        args.map {
+            // TODO what if it.type is null?
+            env.wrap(it.type!!)
+        }
+    }
+    override val typeName: TypeName by lazy {
+        ksType?.typeName() ?: ksTypeReference?.typeName() ?: UNDEFINED
+    }
+    override val nullability by lazy {
+        when (ksType?.nullability) {
+            Nullability.NULLABLE -> XNullability.NULLABLE
+            Nullability.NOT_NULL -> XNullability.NONNULL
+            else -> XNullability.UNKNOWN
+        }
+    }
+
+    override fun asTypeElement(): XTypeElement {
+        TODO("elements are not implemented yet")
+    }
+
+    override fun isAssignableFrom(other: XType): Boolean {
+        check(other is KspType)
+        return ksType.isAssignableFromWithErrorWorkaround(other.ksType)
+    }
+
+    override fun isError(): Boolean {
+        return ksType == null || ksType!!.isError
+    }
+
+    override fun defaultValue(): String {
+        val type = ksType ?: return "null"
+        // NOTE: this does not match the java implementation though it is probably more correct for
+        // kotlin.
+        if (type.nullability == Nullability.NULLABLE) {
+            return "null"
+        }
+        val builtIns = env.resolver.builtIns
+        return when (type) {
+            builtIns.booleanType -> "false"
+            builtIns.byteType, builtIns.shortType, builtIns.intType, builtIns.longType, builtIns
+                .charType -> "0"
+            builtIns.floatType -> "0f"
+            builtIns.doubleType -> "0.0"
+            else -> "null"
+        }
+    }
+
+    override fun boxed(): XType {
+        return this
+    }
+
+    override fun isInt(): Boolean {
+        return ksType.isAssignableFromWithErrorWorkaround(env.resolver.builtIns.intType)
+    }
+
+    override fun isLong(): Boolean {
+        return ksType.isAssignableFromWithErrorWorkaround(env.resolver.builtIns.longType)
+    }
+
+    override fun isByte(): Boolean {
+        return ksType.isAssignableFromWithErrorWorkaround(env.resolver.builtIns.byteType)
+    }
+
+    override fun isNone(): Boolean {
+        return ksType == null
+    }
+
+    override fun isType(): Boolean {
+        return ksType != null && ksType!!.declaration is KSClassDeclaration
+    }
+
+    override fun isTypeOf(other: KClass<*>): Boolean {
+        // closest to what MoreTypes#isTypeOf does.
+        return rawType.typeName.toString() == other.qualifiedName
+    }
+
+    override fun isSameType(other: XType): Boolean {
+        check(other is KspType)
+        // NOTE: this is inconsistent with java where nullability is ignored.
+        // it is intentional but might be reversed if it happens to break use cases.
+        return ksType != null && ksType == other.ksType
+    }
+
+    override fun extendsBound(): XType? {
+        // NOTE: wildcard does not fully exist in kotlin and when we resolve, it always seems to
```

Comment:
extendsBound should be the upper bound of a generic type per naming?

Reply:
this only happens in wildcard but i couldn't create a test case where we would even receive a wildcard representation.
From what i understand [here](https://kotlinlang.org/docs/reference/generics.html), wildcards don't really exists in kotlin and each time i tried `T : Foo`, it always resolved as `Foo`.